lib/galaxy/tools/parameters/basic.py

```diff
@@ -562,8 +562,24 @@ class BooleanToolParameter(ToolParameter):
     def __init__(self, tool, input_source):
         input_source = ensure_input_source(input_source)
         super().__init__(tool, input_source)
-        self.truevalue = input_source.get("truevalue", "true")
-        self.falsevalue = input_source.get("falsevalue", "false")
+        truevalue = input_source.get("truevalue", "true")
+        falsevalue = input_source.get("falsevalue", "false")
+        if tool and tool.profile >= 23.1:
+            if truevalue == falsevalue:
+                raise ParameterValueError("Cannot set true and false to the same value", self.name)
+            if truevalue.lower() == "false":
+                raise ParameterValueError(
+                    f"Cannot set truevalue to [{truevalue}], Galaxy state may encounter issues distinguishing booleans and strings in this case.",
+                    self.name,
+                )
+            if falsevalue.lower() == "true":
+                raise ParameterValueError(
+                    f"Cannot set falsevalue to [{falsevalue}], Galaxy state may encounter issues distinguishing booleans and strings in this case.",
```

Comment:
Wondering why Galaxy has no problem with `falsevalue="false"`?

Reply:
Because it is false either way you interpret it I think.